File: algo-method.com/tasks/1101/main.ts
import { readFileSync } from "fs";
let inputLinesIndex = 0;
const inputLines = readFileSync("/dev/stdin", "utf8").split("\n");
const input = () => inputLines[inputLinesIndex++];

const accumulateMax = (array: number[]) => {
  let now = -Infinity;
  return array.map((v) => (now = Math.max(now, v)));
};

const main = () => {
  const N = +input();
  const A = input().split(" ").map(Number);

  const ans = accumulateMax(A);
  console.log(ans.join("\n"));
};

main();

File: algo-method.com/tasks/1101/main.hs
main :: IO ()
main = do
  n <- getLine
  a <- map read . words <$> getLine

  let ans = accumulateMax a
  mapM_ print ans

accumulateMax :: [Int] -> [Int]
accumulateMax [] = []
accumulateMax [x] = [x]
accumulateMax (x : y : xs) = x : accumulateMax (max x y : xs)
